File: Bakjoon-SWEA/Python/Simulation/12100.py
#import sys
#sys.setrecursionlimit(10 ** 9)
from copy import deepcopy

# 0. 입력 준비
n = int(input())
board = [list(map(int, input().split())) for _ in range(n)]

answer = 0

# 90도를 돌려가며 한 방향으로만 움직임을 구현할 것임.
def rotate(board):
    rotated_board = [[0] * n for _ in range(n)]
    for c in range(n):
        for r in range(n):
            rotated_board[c][r] = board[n - r - 1][c]
    return rotated_board


# 움직임을 수행하는 함수
def move(board):
    moved_board = [[0] * n for _ in range(n)] 
    # 좌측으로 밀어오는 함수를 구현할 건데
    # flag는 해당 수를 moved_board에 이미 올려뒀다는 뜻이고
    # idx을 중심으로 하여 target을 움직여가며 더할 수가 있는지 확인
    for r in range(n):
        row = board[r]
        m_row = moved_board[r]
        
        flag = False
        target = -1
        for idx in range(n):
            # 0이라면 다음 idx로 넘어가야한다.
            if row[idx] == 0:
                continue
            # flag가 있는 상황에서 같은 수가 들어왔다면
            if flag == True and row[idx] == m_row[target]:
                # 값을 2배, target 그대로, flag 회수
                m_row[target] *= 2
                flag = False
            # 다른 상황에서는 0이 아닌 다른 값일 때이므로 flag 세우기
            else:
                target += 1
                m_row[target] = row[idx]
                flag = True
                    
    return moved_board
            


# DFS로 다녀왔다 / 롤백했다를 반복
def DFS(board, cnt):
    # 5번의 이동이 모두 수행되었다면
    if cnt == 5:
        global answer
        for i in range(n):
            answer = max(max(board[i]), answer)
        return
    
    # 4군데 각 방향으로 진행
    for _ in range(4):
        # 원래 상황을 얕은 복사 및 돌리기를 수행한다.
        tmp_board = rotate(deepcopy(board))
        # 복사해둔 것으로 이동을 수행한다
        tmp_board = move(tmp_board)
        # 이동이 수행되었다면 다음 이동을 준비한다.
        DFS(tmp_board, cnt + 1)
        # 원상복구한 후 현재 단계에서의 다른 방향으로의 이동을 시도한다.
        # 맨 위에서 얕은 복사를 수행하므로 복구할 필요 없을 듯.
        
DFS(board, 0)
print(answer)

File: Bakjoon-SWEA/C:C++/Simulation/12100.cpp
#include <iostream>
using namespace std;
const int MAX = 21;
int n = 0;

void DFS(int board[][MAX], int cnt){
    // 들어온 보드를 복사하면서
    // 4방향으로 90도 씩 시계 방향으로 돌거임.
    int rotated_board[MAX][MAX];
    for (int i = 0; i < 4; i++){
        for (int c = 0; c < n; c++){
            for (int r = 0; r < n; r++){
                rotated_board[r][n - c - 1] = board[c][r];
            }
        }

        // 돌린 배열을 바탕으로 숫자 합이 진행된 새로운 배열 생성.
        int after_board[MAX][MAX];
        for (int c = 0; c < n; c++){
            // 열을 하나씩 가져와서 작업할 거임.
            // 1. 돌면서 0이 있는지 파악. 있다면 0을 없애면서 오른쪽으로 붙여야 함.
            // idx -> after_board에 넣을 위치 / target -> rotated_board에서 확인 중인 위치
            int idx = n - 1;
            for (int target = n - 1; target > -1; target--){
                if (rotated_board[c][target] != 0){
                    after_board[c][idx] = rotated_board[c][target];
                    idx++;
                } else {
                    continue;
                }
            }
            // 2. 돌면서 합칠 수 있는게 있는지 파악
            // 한 번 합쳐진 것이 또 합쳐지지 않도록 주의. (4 2 2) -> (4 0 4) -> (0 4 4) -x-> (0 0 8)

        }

    }
}

int main(void){
    cin >> n;

    // 5번 돌려서 최대 값인걸 찾으려면 DFS인데
    // 배열의 맨 첫 포인터를 기준으로 넘기기 때문에
    // 배열은 정해져있는 상수값으로 이어져있어야 함.
    int board[MAX][MAX];

    DFS(board, 0);

}